### DISCUSSION TITLE
tRPC as default before 1.0?

### DISCUSSION THREAD

**Discussion by lkj4** [sveltejs/kit#5242] [Ideas]
__tl;dr: tRPC might be too new and too much of a breaking change to make it the default but it might be also an opportunity and should be discussed before we reach 1.0__

SvelteKit did many right design choices re REST its implementation (shadow endpoints, same router/nice integration with page endpoints, etc.) but I find REST still exhausting and all work-arounds re type safety and DRY stay work-arounds. GraphQL is no option either because it adds bloat, decision fatigue, code gen or other non-DRY means and—the biggest bummer—another type system while TS has perfectly its own. Hence, tRPC might hit the sweet spot here.

tRPC has minimal integration efforts providing full native type safety and the opportunity to detach API routing from the folder structure which is intuitive but cumbersome to refactor and to deal with IRL.

I don't want to debate the pros and cons of REST, GraphQL and tRPC (which we can't ignore entirely) but rather question if REST is really the right default for v1.0.

**Comment by lkj4** [comment]
Three major pros over REST and GraphQL:

- Obviously type safety and type hints _(without codegen!)_, so once you entered `client.queries.getUser(...`, your editor shows...
```
getUser(input: { email: string; }, opts?: ProcedureOptions | undefined): Promise<{ email: string; name: string; }>
```
- Refactoring is now possible, just press `F2` and you change `.getUser` in e.g. `.usersShow` _everywhere_; automatic refactoring is not just impossible with REST and GraphQL but manual changing everything is also a major effort because, e.g., with SvelteKit/REST you need also to change or move folder and file names, respective import statements and so on, not DRY and just a big hassle
- Much, much easier to integrate than GraphQL and zero bloat

If you want to play around yourself check https://stackblitz.com/github/trpc/trpc/tree/next/examples/standalone-server?file=src%2Fserver.ts,src%2Fclient.ts&view=editor and more info on the latest alpha of trpc which also has refactoring: https://github.com/trpc/trpc/blob/next/.tmp/v10-docs.md

**Comment by ivan-kleshnin** [comment]
> tRPC has minimal integration efforts providing full native type safety and the opportunity to detach API routing from the folder structure which is intuitive but cumbersome to refactor and to deal with IRL.

tRPC is a bulky framework with overcomplicated API. Something like integrated Zodios would be much nicer IMO.
